File: Arduino_MEGA_1.54/epd_1inch54.cpp
#include <SPI.h>
#include <string.h>
#include <stdlib.h>		

#include "epd_1inch54.h"
#include "epd_gui.h"

//Configure the serial port to use the standard printf function
//start
int serial_putc( char c, struct __file * )
{
    Serial.write( c );
    return c;
}
void printf_begin(void)
{
    fdevopen( &serial_putc, 0 );
}
//end
   
///////////////////EXTERNAL FUNCTION////////////////////////////////////////////////////////////////////////
/////////////////////delay//////////////////////////////////////
void driver_delay_us(unsigned int xus)  //1us
{
    for(;xus>1;xus--);
}
void driver_delay_xms(unsigned long xms) //1ms
{  
    unsigned long i = 0 , j=0;

    for(j=0;j<xms;j++)
    {
        for(i=0; i<256; i++);
    }
}
void DELAY_S(unsigned int delaytime)     
{
    int i,j,k;
    for(i=0;i<delaytime;i++)
    {
        for(j=0;j<4000;j++)           
        {
            for(k=0;k<222;k++);               
        }
    }
}
//////////////////////SPI///////////////////////////////////
void SPI_Delay(unsigned char xrate)
{
  unsigned char i;
  while(xrate)
  {
      for(i=0;i<2;i++);
      xrate--;
  }
}
void SPI_Write(unsigned char value)                                    
{                                                           
    unsigned char i;  
    SPI_Delay(1);
    for(i=0; i<8; i++)   
    {
        EPD_W21_CLK_0;
        SPI_Delay(1);
        if(value & 0x80)
           EPD_W21_MOSI_1;
        else
           EPD_W21_MOSI_0;   
        value = (value << 1); 
        SPI_Delay(1);
        driver_delay_us(1);
        EPD_W21_CLK_1; 
        SPI_Delay(1);
    }
}

void Epaper_Write_Command(unsigned char command)
{
   SPI_Delay(1);
   EPD_W21_CS_0;                   
   EPD_W21_DC_0;   // command write
   SPI_Write(command);
   EPD_W21_CS_1;
}
void Epaper_Write_Data(unsigned char data)
{
   SPI_Delay(1);
   EPD_W21_CS_0;                   
   EPD_W21_DC_1;   // command write
   SPI_Write(data);
   EPD_W21_CS_1;
}

/////////////////EPD settings Functions/////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////
void EPD_HW_Init(void)
{
  EPD_W21_RST_0;     
  delay(1); 
  EPD_W21_RST_1; //hard reset  
  delay(1); 

  Epaper_READBUSY();   
  Epaper_Write_Command(0x12);  //SWRESET
  Epaper_READBUSY();   
    
  Epaper_Write_Command(0x01); //Driver output control      
  Epaper_Write_Data(0xC7);
  Epaper_Write_Data(0x00);
  Epaper_Write_Data(0x00);

  Epaper_Write_Command(0x11); //data entry mode       
  Epaper_Write_Data(0x01);

  Epaper_Write_Command(0x44); //set Ram-X address start/end position   
  Epaper_Write_Data(0x00);
  Epaper_Write_Data(0x18);    //0x18-->(24+1)*8=200

  Epaper_Write_Command(0x45); //set Ram-Y address start/end position          
  Epaper_Write_Data(0xC7);   //0xC7-->(199+1)=200
  Epaper_Write_Data(0x00);
  Epaper_Write_Data(0x00);
  Epaper_Write_Data(0x00); 

  Epaper_Write_Command(0x3C); //BorderWavefrom
  Epaper_Write_Data(0x05);  
      
  Epaper_Write_Command(0x18); 
  Epaper_Write_Data(0x80);  

  Epaper_Write_Command(0x4E);   // set RAM x address count to 0;
  Epaper_Write_Data(0x00);
  Epaper_Write_Command(0x4F);   // set RAM y address count to 0X199;    
  Epaper_Write_Data(0xC7);
  Epaper_Write_Data(0x00);
  Epaper_READBUSY();
  
}
//////////////////////////////All screen update////////////////////////////////////////////
void EPD_WhiteScreen_ALL(const unsigned char * BWdatas,const unsigned char * RWdatas)
{
   unsigned int i;
   Epaper_Write_Command(0x24);   //write RAM for black(0)/white (1)
   for(i=0;i<ALLSCREEN_GRAGHBYTES;i++)
   {               
     Epaper_Write_Data(pgm_read_byte(&BWdatas[i]));
   }
   Epaper_Write_Command(0x26);   //write RAM for red(0)/white (1)
   for(i=0;i<ALLSCREEN_GRAGHBYTES;i++)
   {               
     Epaper_Write_Data(~pgm_read_byte(&RWdatas[i]));
   }
   EPD_Update();   
}

/////////////////////////////////////////////////////////////////////////////////////////
void EPD_Update(void)
{   
  Epaper_Write_Command(0x22); 
  Epaper_Write_Data(0xF7);   
  Epaper_Write_Command(0x20); 
  Epaper_READBUSY();  

}

void EPD_DeepSleep(void)
{  
  Epaper_Write_Command(0x10); //enter deep sleep
  Epaper_Write_Data(0x01); 
  delay(100);
}
void Epaper_READBUSY(void)
{ 
  while(1)
  {   //=1 BUSY
     if(isEPD_W21_BUSY==0) break;
  }  
}

void EPD_HW_Init_GUI(void)
{
  EPD_W21_RST_0;  // Module reset   
  delay(10);//At least 10ms delay 
  EPD_W21_RST_1;
  delay(10); //At least 10ms delay 
  
  Epaper_READBUSY();   
  Epaper_Write_Command(0x12);  //SWRESET
  Epaper_READBUSY();   
    
  Epaper_Write_Command(0x01); //Driver output control 
  delay(10); //At least 10ms delay  
  Epaper_Write_Data(0xC7);
  delay(10); //At least 10ms delay 
  Epaper_Write_Data(0x00);
  delay(10); //At least 10ms delay 
  Epaper_Write_Data(0x01);//0x00:Show normal 0x01:Show mirror

  Epaper_Write_Command(0x11); //data entry mode       
  Epaper_Write_Data(0x01);

  Epaper_Write_Command(0x44); //set Ram-X address start/end position   
  Epaper_Write_Data(0x00);
  Epaper_Write_Data(0x18);    //0x18-->(24+1)*8=200

  Epaper_Write_Command(0x45); //set Ram-Y address start/end position          
  Epaper_Write_Data(0xC7);    //0xC7-->(199+1)=200
  Epaper_Write_Data(0x00);
  Epaper_Write_Data(0x00);
  Epaper_Write_Data(0x00); 

  Epaper_Write_Command(0x3C); //BorderWavefrom
  Epaper_Write_Data(0x05);  
      
  Epaper_Write_Command(0x18); //Read built-in temperature sensor
  Epaper_Write_Data(0x80);  

  Epaper_Write_Command(0x4E);   // set RAM x address count to 0;
  Epaper_Write_Data(0x00);
  Epaper_Write_Command(0x4F);   // set RAM y address count to 0X199;    
  Epaper_Write_Data(0xC7);
  Epaper_Write_Data(0x00);
  Epaper_READBUSY();
  
}



/////////////////////////////////Single display////////////////////////////////////////////////
// void EPD_WhiteScreen_Black(void)

// {
//     unsigned int i,k;
//     Epaper_Write_Command(0x24);   //write RAM for black(0)/white (1)
//     for(k=0;k<200;k++)
//     {
//         for(i=0;i<25;i++)
//         {
//             Epaper_Write_Data(0x00);
//         }
//     }
//     EPD_Update();
// }

void EPD_WhiteScreen_White(void)

{
    unsigned int i,k;
    Epaper_Write_Command(0x24);   //write RAM for black(0)/white (1)
    for(k=0;k<200;k++)
    {
        for(i=0;i<25;i++)
        {
           Epaper_Write_Data(0xff);
        }
    }
    Epaper_Write_Command(0x26);  
    for(k=0;k<200;k++)
    {
        for(i=0;i<25;i++)
        {
           Epaper_Write_Data(0x00);
        }
    }
    EPD_Update();
}

void EPD_Display(unsigned char *BWImage,unsigned char *RWImage)
{
    unsigned int width, Height,i,j;
    width = (EPD_WIDTH % 8 == 0)? (EPD_WIDTH / 8 ): (EPD_WIDTH / 8 + 1);
    Height = EPD_HEIGHT;

    Epaper_Write_Command(0x24);
    for ( j = 0; j < Height; j++) 
    {
        for ( i = 0; i < width; i++) 
        {
           Epaper_Write_Data(BWImage[i + j * width]);
        }
    }
    Epaper_Write_Command(0x26);
    for ( j = 0; j < Height; j++) 
    {
        for ( i = 0; i < width; i++) 
        {
           Epaper_Write_Data(~RWImage[i + j * width]);
        }
    }
    EPD_Update();    
}


//////////////////////////////////END//////////////////////////////////////////////////

File: raspberry-pi/lgpio/python/epd_1inch54.py
# Seengreat 1.54 Inch E-Paper Display demo
# Author(s):Andy Li from Seengreat

import os
import sys
import spidev
import time
from gpiozero import *
import numpy as np
"""the following pin definiting use BCM"""
PIN_CS     = 8
PIN_DC     = 25
PIN_BUSY   = 24
PIN_RST    = 17
EPD_WIDTH  = 200
EPD_HEIGHT = 200

class EPD_1Inch54():
    def __init__(self):

        # spi init
        self.bus = 0
        self.dev = 0
        self.spi_speed = 8000000
        self.spi = spidev.SpiDev()
        self.spi.open(self.bus, self.dev)
        self.spi.max_speed_hz = self.spi_speed
        self.spi.mode = 0b00
        #self.spi.no_cs = False
        # gpio init
        #self.cs = LED(PIN_CS)
        self.dc = DigitalOutputDevice(PIN_DC,active_high = True,initial_value =False)
        self.rst = DigitalOutputDevice(PIN_RST,active_high = True,initial_value =False)
        self.busy = DigitalInputDevice(PIN_BUSY,pull_up=True,active_state=None)
        self.w = EPD_WIDTH
        self.h = EPD_HEIGHT
        
    def write_cmd(self, cmd):
        """write command"""
        self.dc.off()#wpi.digitalWrite(PIN_DC,wpi.LOW)
        self.spi.writebytes([cmd])
        
    def write_data(self, value):
        """write data"""
        self.dc.on()#wpi.digitalWrite(PIN_DC,wpi.HIGH)
        self.spi.writebytes([value])
        
    def chkstatus(self):
        while self.busy.value==0:
            pass
        
    def reset(self):
        """reset the epd"""
        self.rst.off()#wpi.digitalWrite(PIN_RST,wpi.LOW)
        time.sleep(0.1)
        self.rst.on()#wpi.digitalWrite(PIN_RST,wpi.HIGH)
        time.sleep(0.1)
        
    def hw_init(self):
        """epd init..."""
        self.reset()
        self.chkstatus()
        self.write_cmd(0x12)
        self.chkstatus()
        
        self.write_cmd(0x01) #Driver output control
        self.write_data(0xC7) #199+1=200
        self.write_data(0x00)
        self.write_data(0x00)

        self.write_cmd(0x11) #data entry mode
        self.write_data(0x01)
        
        self.write_cmd(0x44) #set Ram-X address start/end position
        self.write_data(0x00)
        self.write_data(0x18) #0x0F-->(24+1)*8=200
        
        self.write_cmd(0x45) #set Ram-Y address start/end position
        self.write_data(0xC7)
        self.write_data(0x00)
        self.write_data(0x00)
        self.write_data(0x00)
        
        self.write_cmd(0x3C) #BorderWavefrom,
        self.write_data(0x05)

        self.write_cmd(0x21)
        self.write_data(0x00)
        
        self.write_cmd(0x18)
        self.write_data(0x80)
        
        self.write_cmd(0x4E)
        self.write_data(0x00)
        self.write_cmd(0x4F)
        self.write_data(0xC7)
        self.write_data(0x00)
        self.chkstatus()

    def hw_init_gui(self):
        self.reset()
        self.chkstatus()
        self.write_cmd(0x12)  # SWRESET
        self.chkstatus()
        self.write_cmd(0x01)  # Driver output control 
        self.write_data(0xC7)
        self.write_data(0x00)
        self.write_data(0x01)  # 0x00:Show normal 0x01:Show mirror

        self.write_cmd(0x11)  # data entry mode       
        self.write_data(0x01)
        
        self.write_cmd(0x44)  # set Ram-X address start/end position   
        self.write_data(0x00)
        self.write_data(0x18)  # 0x0F-->(24+1)*8=200

        self.write_cmd(0x45)  # set Ram-Y address start/end position        
        self.write_data(0xC7)  # 0xC7-->(199+1)=200
        self.write_data(0x00)
        self.write_data(0x00)
        self.write_data(0x00)

        self.write_cmd(0x3C)  # BorderWavefrom
        self.write_data(0x05)
        
        self.write_cmd(0x21)  # BorderWavefrom
        self.write_data(0x00)

        
        self.write_cmd(0x18)  # Read built-in temperature sensor
        self.write_data(0x80)

        self.write_cmd(0x4E)  # set RAM x address count to 0
        self.write_data(0x00)
        self.write_cmd(0x4F) # set RAM y address count to 0X199
        self.write_data(0xC7)
        self.write_data(0x00)
        self.chkstatus()
        
    def update(self):
        self.write_cmd(0x22)
        self.write_data(0xF7)
        self.write_cmd(0x20)
        self.chkstatus()

    # display
    def whitescreen_all(self,BWdatas,RWdatas):
        self.write_cmd(0x24) #write RAM for black(0)/white (1)
        for i in range(5000):
            self.write_data(BWdatas[i])
        self.write_cmd(0x26) #write RAM for red(0)/white (1)
        for i in range(5000):
            self.write_data(~RWdatas[i])
        self.update()

    def whitescreen_white(self):
        self.write_cmd(0x24) # write RAM for black(0) / white(1)
        for k in range(200):
            for i in range(25):
                self.write_data(0xff)
        self.write_cmd(0x26) 
        for k in range(200):
            for i in range(25):
                self.write_data(0x00)
        self.update()

    def sleep(self):
        self.write_cmd(0x10)
        self.write_data(0x01)
        time.sleep(0.01)

    def display_clear(self):
        self.write_cmd(0x24)
        for i in range(5000):
            self.write_data(0xFF)
        self.write_cmd(0x26)
        for i in range(5000):
            self.write_data(0x00)
        self.update()

    def display(self, BWimage,RWimage):
        if self.w%8 == 0:
            width = self.w//8
        else:
            width = self.w//8+1
        height = self.h

        self.write_cmd(0x24)
        for j in range(height):
            for i in range(width):
               self.write_data(BWimage[i + j * width])
        self.write_cmd(0x26)
        for j in range(height):
            for i in range(width):
               self.write_data(~RWimage[i + j * width])
        self.update()
